[Dx 963] Add do_not_track Analytics FAQ (#4032)
TykTechnologies/tyk-docs: tyk-docs/content/developer-support/frequently-asked-questions/what_is_the_performance_impact_of_analytics.md

---
title: "What Is The Performance Impact Of Analytics?"
date: 2024-01-22
tags: ["do_not_track", "Analytics", "RPS", "Requests Per Second", "CPU", "high load", "high traffic"]
description: "This FAQ explains how analytics impacts system performance and how to disable using do_not_track"
---

> Tyk allows analytics to be recorded and stored in backend storage for API endpoints, via [Tyk Pump]({{< ref "tyk-stack/tyk-pump/tyk-analytics-record-fields" >}}). Example details include: the originating host, API version, request method, request path etc. This can introduce the problem of increased CPU load and reduced RPS (requests per second).

Reviewer: ```suggestion
Tyk allows analytics to be recorded and stored in backend storage for API endpoints, via [Tyk Pump]({{< ref "tyk-stack/tyk-pump/tyk-analytics-record-fields" >}}). 

Tyk allows you to keep track of how your APIs are used by recording the traffic it's processing and in the background, sending the analytics data to Redis and then, using *Tyk Pump* a [data sink]{{<ref "tyk-stack/tyk-pump/other-data-stores">}} of you choice. The [analytics data includes]({{< ref "tyk-stack/tyk-pump/tyk-analytics-record-fields" >}})  details like originating host (where the request is coming from), which Tyk API version was used, HTTP method request, request path etc. This recording can sometimes introduce the problem of increased CPU load and reduced RPS (requests per second).
```

Reviewer: We should also explain what is going to happen on this page 
"We are going to share with you config to easy this as well as an example that demonstrates this field ...."

Reviewer 2: I would avoid the use of the word "track" except when referring to the middleware (to minimise the risk of confusion)
I recommend we talk about how the Tyk Gateway generates "transaction records" or "analytics records" (where a transaction is an API request + response).
Then the Tyk Pump transfers the raw records to persistent storage (Mongo/SQL)
The Tyk Pump can also create aggregated data (using different data keys - API ID, access key, endpoint, response status code, location) and stores those in persistent storage
...and finally Tyk Dashboard can retrieve those data and provide analytics based on the aggregated data; it also provides the log browser that gives visibility to the raw transaction logs.

Notes:
 - by default the Gateway will generate transaction records for all APIs
 - you can selectively disable this using `do-not-track`: if this is employed then the Gateway does not generate transaction records for that API/endpoint, so there is nothing for Pump to pick up
 - the `track` middleware ONLY affects the "endpoint popularity" aggregated data generated by Pump

Author: updated

> {{< img src="/img/faq/enable-multiple-analytics-keys/redis_single.png" alt="High CPU load" width="864px" >}}

Reviewer: I don't think this image is relevant to this problem

Author: removed

> Tyk provides the [do_not_track]({{<ref "advanced-configuration/transform-traffic/endpoint-designer#do-not-track-endpoint">}}) configuration parameter to disable analytics recording at two levels:

Reviewer: Before introducing this field, we should explain it first. something like
"...record less info can reduce the load. Tyk allows you to decide which info is important and should be recorded and which can be skipped saving CPU cycles and storage...

Reviewer: @dcs3spp

Author: updated

- **Per API**: Tyk Gateway will not record analytics for requests/responses of all the endpoints of this API.
- **Endpoint**: Tyk Gateway will not record analytics for requests to the selected endpoint. Other endpoints will be tracked as normal.

Enabling [do_not_track]({{<ref "advanced-configuration/transform-traffic/endpoint-designer#do-not-track-endpoint">}}) can help reduce analytics traffic for systems that are under high load, exhibiting high traffic, e.g. social media platforms, streaming services and financial services and trading platforms.

Application decisions need to be made concerning which endpoints are non critical and can thus have analytics disabled. Furthermore, benchmarking and testing will be required to evaluate the actual benefits for the application specific use case.

> In the *Tyk Dashboard API* screen below, there are two APIs, *track* and *notrack*. The APIs were created to conduct a simple load test, to show the gateway's RPS (requests per second) for each API:

Reviewer 2: I would be inclined to give these APIs slightly different names to avoid potential confusion with the `track` and `do_not_track` middleware / configurations.

Perhaps:
 - trackedAPI
 - untrackedAPI

## What Are Analytics?
Tyk Gateway allows analytics to be recorded and stored in a persistent data store (MongoDB/SQL) for all APIs by default, via [Tyk Pump]({{< ref "tyk-stack/tyk-pump/tyk-analytics-record-fields" >}}).

Tyk Gateway generates transaction records for each API request and response, containing [analytics data]({{< ref "tyk-stack/tyk-pump/tyk-analytics-record-fields" >}}) relating to: the originating host (where the request is coming from), which Tyk API version was used, the HTTP method requested and request path etc.

The transaction records are transmitted to Redis and subsequently transferred to a persistent [data store]({{< ref "tyk-stack/tyk-pump/other-data-stores" >}}) of your choice via Tyk Pump. Furthermore, Tyk Pump can also be configured to [aggregate]({{< ref "tyk-dashboard-analytics#aggregated-analytics" >}}) the transaction records (using different data keys - API ID, access key, endpoint, response status code, location) and write to a persistent data store. Tyk Dashboard uses this data for:

> - Displaying [analytics]({{< ref "tyk-dashboard-analytics" >}}) based on the aggregated data.

Reviewer: ```suggestion
- **Aggregated analytics** - Displaying [analytics]({{< ref "tyk-dashboard-analytics" >}}) based on the aggregated data.
```

> - Browsing logs for the [raw transaction records]({{< ref "tyk-stack/tyk-manager/analytics/log-browser" >}}).

Reviewer: ```suggestion
- [Log Broswer]({{< ref "tyk-stack/tyk-manager/analytics/log-browser" >}}) - Browsing logs for the raw transaction records.
```

## How Do Analytics Impact Performance?

Analytics may introduce the problem of increased CPU load and a decrease in the number of requests per second (RPS).

In the *Tyk Dashboard API* screen below, there are two APIs, *track* and *notrack*. The APIs were created to conduct a simple load test, to show the gateway's RPS (requests per second) for each API:

> - **track**: Analytics are tracked for an API, i.e. *do_not_track* is false.

Reviewer 2: ```suggestion
- **track**: Traffic to this API is tracked, i.e. transaction records are generated for each request/response.
```

> - **notrack**: Analytics are not tracked for an API, i.e. *do_not_track* is true.

Reviewer 2: ```suggestion
- **notrack**: Traffic to this API is not tracked, i.e. transaction records are not generated for each request/response.
```

{{< img src="img/faq/do-not-track-usage-scenario/dashboard_apis_measured.png" alt="apis measured in Tyk Dashboard" width="864">}}

> 100,000 requests were sent to each API and the total number of requests per second was measured. The results for the *tracked* API are displayed in the left pane terminal window; with the right pane showing the results for the *untracked* API.

Reviewer 2: ```suggestion
100,000 requests were sent to each API and the rate at which Tyk was able to handle those requests (number of requests per second) was measured. The results for the *tracked* API are displayed in the left pane terminal window; with the right pane showing the results for the *untracked* API.
```

{{< img src="img/faq/do-not-track-usage-scenario/do_not_track_performance_impact.png" alt="measuring do_not_track API performance impact" >}}

We can see that **19,253.75** RPS was recorded for the *untracked* API; with **16,743.6011** RPS reported for the *tracked* API. The number of requests per second decreased by **13.0372%** when analytics was enabled.

## What Can Be Done To Address This Performance Impact?

Tyk is configurable, allowing fine grained control over which information should be recorded and which can be skipped, thus reducing CPU cycles, traffic and storage.

> Users can [selectively control]({{<ref "product-stack/tyk-gateway/middleware/do-not-track-middleware">}}) the generation of analytics :

Reviewer 2: ```suggestion
Users can [selectively prevent]({{<ref "product-stack/tyk-gateway/middleware/do-not-track-middleware">}}) the generation of analytics:
```

- **Per API**: Tyk Gateway will not create records for requests/responses for any endpoints of an API.
- **Per Endpoint**: Tyk Gateway will not create records for requests/responses for specific endpoints.

> When set, this prevents Tyk Gateway from generating the transaction records. Without transaction records, Tyk Pump will not transfer analytics to the chosen persistent data store. It's worth noting that the ["track" middleware]({{< ref "product-stack/tyk-dashboard/advanced-configurations/analytics/activity-by-endpoint" >}}) exclusively influences the generation of *endpoint popularity* aggregated data by Tyk Pump.

Reviewer: you can't add " inside a relative link, like this:
> ["track" middleware]
Check the preview and see it's not working. We need CI to find it

Author: thanks @letzya I've removed the surrounding quotes

In the *Tyk Dashboard API* screen below, there are two APIs, *track* and *notrack*. The APIs were created to conduct a simple load test, to show the gateway's RPS (requests per second) for each API:

- **track**: Traffic to this API is tracked, i.e. transaction records are generated for each request/response.
- **notrack**: Traffic to this API is not tracked, i.e. transaction records are not generated for each request/response.

{{< img src="img/faq/do-not-track-usage-scenario/dashboard_apis_measured.png" alt="apis measured in Tyk Dashboard" width="864">}}

100,000 requests were sent to each API and the rate at which Tyk was able to handle those requests (number of requests per second) was measured. The results for the *tracked* API are displayed in the left pane terminal window; with the right pane showing the results for the *untracked* API.

### Tracked API Performance

{{< img src="img/faq/do-not-track-usage-scenario/track.png" alt="measuring tracked API performance impact" >}}

### Untracked API Performance

{{< img src="img/faq/do-not-track-usage-scenario/notrack.png" alt="measuring do_not_track API performance impact" >}}

> ### Summary

Reviewer: ```suggestion
### Explaining the results
```

> We can see that **19,253.75** RPS was recorded for the *untracked* API; with **16,743.6011** RPS reported for the *tracked* API. The number of requests per second decreased by **13.0372%** when analytics was enabled.

Reviewer: ```suggestion
We can see that **19,253.75** RPS was recorded for the *untracked* API; with **16,743.6011** RPS reported for the *tracked* API. The number of requests per second decreased by **~13%** when analytics was enabled.
```

description: "This FAQ explains how analytics impacts system performance and how to disable using do_not_track"
---

This FAQ explains how analytics may impact performance relating to high CPU load and reduced Requests Per Second (RPS). Subsequently, it describes how to configure Tyk to reduce this performance impact for platforms that exhibit high volumes of traffic.

## What Are Analytics?

> Tyk Gateway allows analytics to be recorded and stored in backend storage (MongoDB/SQL) for all APIs by default, via [Tyk Pump]({{< ref "tyk-stack/tyk-pump/tyk-analytics-record-fields" >}}).

Reviewer 2: I recommend checking whether we consistently use the term "persistent storage" or "backend storage" or if we use them interchangeably (e.g. https://tyk.io/docs/tyk-dashboard/database-options/)

Author: Thanks @andyo-tyk good spot!! I've updated the FAQ content to use persistent data store

> Tyk Gateway generates transaction records for each API request and response, containing [analytics data]({{< ref "tyk-stack/tyk-pump/tyk-analytics-record-fields" >}}) relating to: the originating host (where the request is coming from), which Tyk API version was used, the HTTP method requested and request path etc.

Reviewer 2: I think that `tyk-analytics-record-fields.md` should probably sit in the Tyk Gateway part of the docs, as (if I understand correctly) it describes the transaction records/traffic logs generated by the Gateway. Yes, Pump picks them up and moves them around, but they are a Gateway artifact. That is, unless the records created by Gateway are actually different and this file describes them post-Pump (if, say, Pump enriches them somehow).

Author: @andyo-tyk Looks like the Pump [analytics fields](https://tyk.io/docs/tyk-stack/tyk-pump/tyk-analytics-record-fields/) are based on raw request/response details. Could raise a separate ticket and then address the scope in a new PR.

Reviewer 2: Yes, don't think it needs to be fixed in this PR, but something to check out.

Reviewer: it's an auto gen pr so check where it's coming from.

Reviewer: I think it's coming from pump so no need to do anything IMO. We say "Tyk Analytics Record" in the pump while in the GW it's part of the API def.

We can see that **19,253.75** RPS was recorded for the *untracked* API; with **16,743.6011** RPS reported for the *tracked* API. The number of requests per second decreased by **13.0372%** when analytics was enabled.

## What Can Be Done To Address This Performance Impact?

Tyk is configurable, allowing fine grained control over which information should be recorded and which can be skipped, thus reducing CPU cycles, traffic and storage.

> Users can selectively disable analytics using the [do_not_track]({{<ref "advanced-configuration/transform-traffic/endpoint-designer#do-not-track-endpoint">}}) configuration parameter. This can configured:

Reviewer 2: With Tyk OAS, we're planning to implement the "per-API" suppression of analytics creation slightly differently, inverting the logic so it is "enable/disable analytics" which seems more intuitive (and then you can selectively choose not to track individual endpoints if analytics is enabled for the API).

So, in anticipation of this I suggest:

```suggestion
Users can [selectively control]({{<ref "product-stack/tyk-gateway/middleware/do-not-track-middleware">}}) the generation of analytics :
```

Author: Ok thanks @andyo-tyk I'll update with suggestion

> - **Per API**: Tyk Gateway will not record analytics for requests/responses of all endpoints of an API.

Reviewer 2: ```suggestion
- **Per API**: Tyk Gateway will not create records for requests/responses for any endpoints of an API.
```

> - **Per Endpoint**: Tyk Gateway will not record analytics for requests for a specific endpoint. Other endpoints will be tracked as normal.

Reviewer 2: ```suggestion
- **Per Endpoint**: Tyk Gateway will not create records for requests/responses for specific endpoints.
```

> When set, this prevents Tyk Gateway from generating the transaction records. Without transaction records, Tyk Pump will not transfer analytics to the chosen data sink. It's worth noting that the "track" middleware exclusively influences the generation of *endpoint popularity* aggregated data by *Tyk Pump*.

Reviewer 2: ```suggestion
When set, this prevents Tyk Gateway from generating the transaction records. Without transaction records, Tyk Pump will not transfer analytics to the chosen data sink. It's worth noting that the ["track" middleware]({{< ref "product-stack/tyk-dashboard/advanced-configurations/analytics/activity-by-endpoint" >}}) exclusively influences the generation of *endpoint popularity* aggregated data by *Tyk Pump*.
```

## Conclusion

> Enabling [do_not_track]({{<ref "advanced-configuration/transform-traffic/endpoint-designer#do-not-track-endpoint">}}) disables recording of analytics per API or endpoint. This helps reduce CPU cycles and network requests for systems that exhibit high load and traffic, e.g. social media platforms, streaming, financial services and trading platforms.

Reviewer 2: ```suggestion
[Disabling]({{<ref "product-stack/tyk-gateway/middleware/do-not-track-middleware">}})  the creation of analytics (either per API or for specific endpoints) helps to reduce CPU cycles and network requests for systems that exhibit high load and traffic, e.g. social media platforms, streaming, financial services and trading platforms.
```
